doc/reference/kernel/other/fatal.rst

```diff
@@ -0,0 +1,245 @@
+the compiler. If the compiler supports it, the provided message will be printed
+to provide further context.
+
+Unlike ``__ASSERT()``, the message must be a static string, without
+``printf()``-like format codes or extra arguments.
+
+For example, suppose this check fails:
+
+.. code-block:: c
+
+	BUILD_ASSERT_MSG(FOO == 2000,
+			 "Invalid value of FOO");
+
+With GCC, the output resembles:
+
+.. code-block:: none
+
+	/home/apboie/projects/zephyr3/zephyr/tests/kernel/fatal/src/main.c: In function 'test_main':
+	/home/apboie/projects/zephyr3/zephyr/include/toolchain/gcc.h:28:37: error: static assertion failed: "Invalid value of FOO"
+	 #define BUILD_ASSERT_MSG(EXPR, MSG) _Static_assert(EXPR, MSG)
+					     ^~~~~~~~~~~~~~
+	/home/apboie/projects/zephyr3/zephyr/tests/kernel/fatal/src/main.c:370:2: note: in expansion of macro 'BUILD_ASSERT_MSG'
+	  BUILD_ASSERT_MSG(FOO == 2000,
+	  ^~~~~~~~~~~~~~~~
+
+BUILD_ASSERT()
+--------------
+
+This works just like ``BUILD_ASSERT_MSG()`` except there is no supplemental
+message provided, and like ``__ASSERT_NO_MSG()`` its use is discouraged.
+
+Kernel Oops
+===========
+
+A kernel oops is a software triggered fatal error invoked by k_oops().
+This should be used to indicate an unrecoverable condition in application
+logic.
+
+The fatal error reason code generated will be ``K_ERR_KERNEL_OOPS``.
+
+Kernel Panic
+============
+
+A kernel arror is a software triggered fatal error invoked by
+:c:func:`k_panic()`.  This should be used to indicate that the Zephyr kernel is
+in an unrecoverable state. Implementations of
+:c:func:`k_sys_fatal_error_handler()` should not return if the kernel
+encounters a panic condition, the entire system needs to be reset.
+
+Threads running in user mode are not permitted to invoke :c:func:`k_panic()`,
+and doing so will generate a kernel oops instead. Otherwise, the fatal error
+reason code generated will be ``K_ERR_KERNEL_PANIC``.
+
+Exceptions
+**********
+
+Spurious Interrupts
+===================
+
+If the CPU receives a hardware interrupt on an interrupt line that has not
+had a handler installed with ``IRQ_CONNECT()`` or ``irq_connect_dynamic()``,
+then the kernel will generate a fatal error with the readon code
+``K_ERR_SPURIOUS_IRQ()``.
+
+Stack Overflows
+===============
+
+In the event that a thread pushes more data onto its execution stack than its
+stack buffer provides, the kernel may be able to detect this situation and
+generate a fatal error with a reason code of ``K_ERR_STACK_CHK_FAIL``.
+
+If a thread is running in user mode, then stack overflows are always caught,
+as the thread will simply not have permission to write to adjacent memory
+addresses outside of the stack buffer. Because this is enforced by the
+memory protection hardware, there is no risk of data corruption to memory
+that the thread would not otherwise be able to write to.
+
+If a thead is running in supervisor mode, or if ``CONFIG_USERSPACE`` is not
```

Comment:
I would reverse this order.
I.e. "If CONFIG_USERSPACE is not enabled, or if it is enabled but the thread is running in supervisor mode"

Reply:
When referencing config options you can use this syntax to link to that option's documentation:
```
:option:`CONFIG_USERSPACE`
:option:`CONFIG_HW_STACK_PROTECTION`
:option:`CONFIG_STACK_SENTINEL`
```